docs/proposals/cluster-capacity.md

```diff
@@ -0,0 +1,446 @@
+The binding process delegates pod to [client](../../plugin/pkg/scheduler/factory/factory.go#L336).
+Thus, it is simulated as a part of the recomputation.
+
+As the Kubelet uses scheduler's [NodeInfo.addPod](../../plugin/pkg/scheduler/schedulercache/node_info.go#L171) to compute the overall consumption of resources, the recomputation can reuse the same code.
+
+Before the prediction starts, all the predictor caches need to be populated.
+By default, every scheduler configuration is built from the [config factory](../../plugin/pkg/scheduler/factory/factory.go#L100).
+The factory provides various ways how to build the configuration.
+Building process also [initialize](../../plugin/pkg/scheduler/factory/factory.go#L387) reflectors and informers responsible for caches population and updates.
+Each reflector and informer has its own cache and use exactly one of the following functions to create ListerWatcher:
+
+* `createUnassignedNonTerminatedPodLW`
+* `createAssignedNonTerminatedPodLW`
+* `createNodeLW`
+* `createPersistentVolumeLW`
+* `createPersistentVolumeClaimLW`
+* `createServiceLW`
+* `createControllerLW`
+* `createReplicaSetLW`
+
+Thus, when implementing the local client, all the caches need to be populated with the current state
+using exactly the same ListerWatchers per predictor cache.
+Once the caches are populated, all the reflectors and informers are terminated.
+Thus, the local client can return content of the caches to simulate communication with the Apiserver.
+
+At this point the factory can build the configuration pointing to the local client.
+Scheduler gets created and the prediction can start.
+At the end of each scheduling iteration the binding process is converted to cache update.
+Normally, each binding results in watch event once the pod gets run on kubelet and node status gets reflected in the Apiserver.
+Thus, as a part of updating the predictor caches, watch event is generated (e.g. in the [RestClient](../../pkg/watch/streamwatcher.go#L114) it corresponds to sending an item to a channel).
+
+From the point of view of the scheduler the local client behaves the same way as any client,
+i.e. it provides ``Listers``/``Watchers`` implementing ``List`` and ``Watch`` interface
+
+## Examples
+
+**Expected output** (still very sci-fi, ``kubectl cluster-capacity`` is only illustrative, the command can change):
+
+```
+$ kubectl cluster-capacity -f pod.json
+PodRequirements:
+- cpu: 2.5
+- memory: 40Mi
+- PVClaim:
+  - count: 2
+  - total_storage: 10Gi
+- labels:
+  - name: zone
+    value: Europe
+  - name: type
+    value: HPC
+
+The cluster can schedule 23 instance(s) of the pod.
+```
+
+Optionaly, list of nodes individual pods get scheduled on can be returned.
+
+## Roadmap
+
+1. implement the local client, prediction framework, REST API, limit considered resources for cpu, memory and gpu
+1. extend considered resources, extend prediction
+
+## Future
+
+* Take re-scheduler into account with possible de-fragmentation
```

Comment:
Does this mean the rescheduler could propose compaction scenarios to the capacity service in order to plan the least-disruptive shuffling necessary to make space for some set of pending pods?

If so, would that affect the API in any way that you'd want to capture here, e.g. optionally supplying a starting cluster state as part of the query?